// react-spring/hooks.js
import _inheritsLoose from '@babel/runtime/helpers/esm/inheritsLoose';
import _assertThisInitialized from '@babel/runtime/helpers/esm/assertThisInitialized';
import _extends from '@babel/runtime/helpers/esm/extends';
import _objectWithoutPropertiesLoose from '@babel/runtime/helpers/esm/objectWithoutPropertiesLoose';
import React from 'react';
import ReactDOM from 'react-dom';

var bugfixes = undefined;
var applyAnimatedValues = undefined;
var colorNames = [];
var requestFrame = function requestFrame(cb) {
  return typeof window !== 'undefined' && window.requestAnimationFrame(cb);
};
var cancelFrame = function cancelFrame(cb) {
  return typeof window !== 'undefined' && window.cancelAnimationFrame(cb);
};
var interpolation = undefined;
var now = function now() {
  return Date.now();
};
var defaultElement = undefined;
var createAnimatedStyle = undefined;
var injectApplyAnimatedValues = function injectApplyAnimatedValues(fn, transform) {
  return applyAnimatedValues = {
    fn: fn,
    transform: transform
  };
};
var injectColorNames = function injectColorNames(names) {
  return colorNames = names;
};
var injectBugfixes = function injectBugfixes(fn) {
  return bugfixes = fn;
};
var injectInterpolation = function injectInterpolation(cls) {
  return interpolation = cls;
};
var injectFrame = function injectFrame(raf, caf) {
  var _ref;

  return _ref = [raf, caf], requestFrame = _ref[0], cancelFrame = _ref[1], _ref;
};
var injectNow = function injectNow(nowFn) {
  return now = nowFn;
};
var injectDefaultElement = function injectDefaultElement(el) {
  return defaultElement = el;
};
var injectCreateAnimatedStyle = function injectCreateAnimatedStyle(factory) {
  return createAnimatedStyle = factory;
};

var Globals = /*#__PURE__*/Object.freeze({
  get bugfixes () { return bugfixes; },
  get applyAnimatedValues () { return applyAnimatedValues; },
  get colorNames () { return colorNames; },
  get requestFrame () { return requestFrame; },
  get cancelFrame () { return cancelFrame; },
  get interpolation () { return interpolation; },
  get now () { return now; },
  get defaultElement () { return defaultElement; },
  get createAnimatedStyle () { return createAnimatedStyle; },
  injectApplyAnimatedValues: injectApplyAnimatedValues,
  injectColorNames: injectColorNames,
  injectBugfixes: injectBugfixes,
  injectInterpolation: injectInterpolation,
  injectFrame: injectFrame,
  injectNow: injectNow,
  injectDefaultElement: injectDefaultElement,
  injectCreateAnimatedStyle: injectCreateAnimatedStyle
});

var Animated =
/*#__PURE__*/
function () {
  function Animated() {}

  var _proto = Animated.prototype;

  _proto.attach = function attach() {};

  _proto.detach = function detach() {};

  _proto.getValue = function getValue() {};

  _proto.getAnimatedValue = function getAnimatedValue() {
    return this.getValue();
  };

  _proto.addChild = function addChild(child) {};

  _proto.removeChild = function removeChild(child) {};

  _proto.getChildren = function getChildren() {
    return [];
  };

  return Animated;
}();

var getValues = function getValues(object) {
  return Object.keys(object).map(function (k) {
    return object[k];
  });
};

var AnimatedWithChildren =
/*#__PURE__*/
function (_Animated) {
  _inheritsLoose(AnimatedWithChildren, _Animated);

  function AnimatedWithChildren() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Animated.call.apply(_Animated, [this].concat(args)) || this;
    _this.children = [];

    _this.getChildren = function () {
      return _this.children;
    };

    _this.getPayload = function (index) {
      if (index === void 0) {
        index = undefined;
      }

      return index !== void 0 && _this.payload ? _this.payload[index] : _this.payload || _assertThisInitialized(_assertThisInitialized(_this));
    };

    return _this;
  }

  var _proto = AnimatedWithChildren.prototype;

  _proto.addChild = function addChild(child) {
    if (this.children.length === 0) this.attach();
    this.children.push(child);
  };

  _proto.removeChild = function removeChild(child) {
    var index = this.children.indexOf(child);

    if (index === -1) {
      if (process.env.NODEENV !== 'production') {
        console.warn("Trying to remove a child that doesn't exist");
      }

      return;
    }

    this.children.splice(index, 1);
    if (this.children.length === 0) this.detach();
  };

  return AnimatedWithChildren;
}(Animated);
var AnimatedArrayWithChildren =
/*#__PURE__*/
function (_AnimatedWithChildren) {
  _inheritsLoose(AnimatedArrayWithChildren, _AnimatedWithChildren);

  function AnimatedArrayWithChildren() {
    var _this2;

    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    _this2 = _AnimatedWithChildren.call.apply(_AnimatedWithChildren, [this].concat(args)) || this;
    _this2.payload = [];

    _this2.getAnimatedValue = function () {
      return _this2.getValue();
    };

    _this2.attach = function () {
      return _this2.payload.forEach(function (p) {
        return p instanceof Animated && p.addChild(_assertThisInitialized(_assertThisInitialized(_this2)));
      });
    };

    _this2.detach = function () {
      return _this2.payload.forEach(function (p) {
        return p instanceof Animated && p.removeChild(_assertThisInitialized(_assertThisInitialized(_this2)));
      });
    };

    return _this2;
  }

  return AnimatedArrayWithChildren;
}(AnimatedWithChildren);
var AnimatedObjectWithChildren =
/*#__PURE__*/
function (_AnimatedWithChildren2) {
  _inheritsLoose(AnimatedObjectWithChildren, _AnimatedWithChildren2);

  function AnimatedObjectWithChildren() {
    var _this3;

    for (var _len3 = arguments.length, args = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
      args[_key3] = arguments[_key3];
    }

    _this3 = _AnimatedWithChildren2.call.apply(_AnimatedWithChildren2, [this].concat(args)) || this;
    _this3.payload = {};

    _this3.getAnimatedValue = function () {
      return _this3.getValue(true);
    };

    _this3.attach = function () {
      return getValues(_this3.payload).forEach(function (s) {
        return s instanceof Animated && s.addChild(_assertThisInitialized(_assertThisInitialized(_this3)));
      });
    };

    _this3.detach = function () {
      return getValues(_this3.payload).forEach(function (s) {
        return s instanceof Animated && s.removeChild(_assertThisInitialized(_assertThisInitialized(_this3)));
      });
    };

    return _this3;
  }

  var _proto2 = AnimatedObjectWithChildren.prototype;

  _proto2.getValue = function getValue(animated) {
    if (animated === void 0) {
      animated = false;
    }

    var payload = {};

    for (var key in this.payload) {
      var value = this.payload[key];
      if (animated && !(value instanceof Animated)) continue;
      payload[key] = value instanceof Animated ? value[animated ? 'getAnimatedValue' : 'getValue']() : value;
    }

    return payload;
  };

  return AnimatedObjectWithChildren;
}(AnimatedWithChildren);

var Interpolation =
/*#__PURE__*/
function () {
  function Interpolation() {}

  Interpolation.create = function create(config, arg) {
    if (typeof config === 'function') return config;else if (interpolation && config.output && typeof config.output[0] === 'string') return interpolation(config);else if (Array.isArray(config)) return Interpolation.create({
      range: config,
      output: arg
    });
    var outputRange = config.output;
    var inputRange = config.range || [0, 1];

    var easing = config.easing || function (t) {
      return t;
    };

    var extrapolateLeft = 'extend';
    var map = config.map;
    if (config.extrapolateLeft !== undefined) extrapolateLeft = config.extrapolateLeft;else if (config.extrapolate !== undefined) extrapolateLeft = config.extrapolate;
    var extrapolateRight = 'extend';
    if (config.extrapolateRight !== undefined) extrapolateRight = config.extrapolateRight;else if (config.extrapolate !== undefined) extrapolateRight = config.extrapolate;
    return function (input) {
      var range = findRange(input, inputRange);
      return interpolate(input, inputRange[range], inputRange[range + 1], outputRange[range], outputRange[range + 1], easing, extrapolateLeft, extrapolateRight, map);
    };
  };

  return Interpolation;
}();

function interpolate(input, inputMin, inputMax, outputMin, outputMax, easing, extrapolateLeft, extrapolateRight, map) {
  var result = map ? map(input) : input; // Extrapolate

  if (result < inputMin) {
    if (extrapolateLeft === 'identity') return result;else if (extrapolateLeft === 'clamp') result = inputMin;
  }

  if (result > inputMax) {
    if (extrapolateRight === 'identity') return result;else if (extrapolateRight === 'clamp') result = inputMax;
  }

  if (outputMin === outputMax) return outputMin;
  if (inputMin === inputMax) return input <= inputMin ? outputMin : outputMax; // Input Range

  if (inputMin === -Infinity) result = -result;else if (inputMax === Infinity) result = result - inputMin;else result = (result - inputMin) / (inputMax - inputMin); // Easing

  result = easing(result); // Output Range

  if (outputMin === -Infinity) result = -result;else if (outputMax === Infinity) result = result + outputMin;else result = result * (outputMax - outputMin) + outputMin;
  return result;
}

function findRange(input, inputRange) {
  for (var i = 1; i < inputRange.length - 1; ++i) {
    if (inputRange[i] >= input) break;
  }

  return i - 1;
}

var AnimatedInterpolation =
/*#__PURE__*/
function (_AnimatedArrayWithChi) {
  _inheritsLoose(AnimatedInterpolation, _AnimatedArrayWithChi);

  function AnimatedInterpolation(parents, _config, _arg) {
    var _this;

    _this = _AnimatedArrayWithChi.call(this) || this;

    _this.getValue = function () {
      var _this2;

      return (_this2 = _this).calc.apply(_this2, _this.payload.map(function (value) {
        return value.getValue();
      }));
    };

    _this.updateConfig = function (config, arg) {
      return _this.calc = Interpolation.create(config, arg);
    };

    _this.interpolate = function (config, arg) {
      return new AnimatedInterpolation(_assertThisInitialized(_assertThisInitialized(_this)), config, arg);
    };

    _this.payload = // AnimatedArrays should unfold, except AnimatedInterpolation which is taken as is
    parents instanceof AnimatedArrayWithChildren && !parents.updateConfig ? parents.payload : Array.isArray(parents) ? parents : [parents];
    _this.calc = Interpolation.create(_config, _arg);
    return _this;
  }

  return AnimatedInterpolation;
}(AnimatedArrayWithChildren);
var interpolate$1 = function interpolate(parents, config, arg) {
  return parents && new AnimatedInterpolation(parents, config, arg);
};

/**
 * Animated works by building a directed acyclic graph of dependencies
 * transparently when you render your Animated components.
 *
 *               new Animated.Value(0)
 *     .interpolate()        .interpolate()    new Animated.Value(1)
 *         opacity               translateY      scale
 *          style                         transform
 *         View#234                         style
 *                                         View#123
 *
 * A) Top Down phase
 * When an Animated.Value is updated, we recursively go down through this
 * graph in order to find leaf nodes: the views that we flag as needing
 * an update.
 *
 * B) Bottom Up phase
 * When a view is flagged as needing an update, we recursively go back up
 * in order to build the new value that it needs. The reason why we need
 * this two-phases process is to deal with composite props such as
 * transform which can receive values from multiple parents.
 */

function findAnimatedStyles(node, styles) {
  if (typeof node.update === 'function') styles.add(node);else node.getChildren().forEach(function (child) {
    return findAnimatedStyles(child, styles);
  });
}
/**
 * Standard value for driving animations.  One `Animated.Value` can drive
 * multiple properties in a synchronized fashion, but can only be driven by one
 * mechanism at a time.  Using a new mechanism (e.g. starting a new animation,
 * or calling `setValue`) will stop any previous ones.
 */


var AnimatedValue =
/*#__PURE__*/
function (_AnimatedWithChildren) {
  _inheritsLoose(AnimatedValue, _AnimatedWithChildren);

  function AnimatedValue(_value) {
    var _this;

    _this = _AnimatedWithChildren.call(this) || this;

    _this.getValue = function () {
      return _this.value;
    };

    _this.updateStyles = function () {
      return findAnimatedStyles(_assertThisInitialized(_assertThisInitialized(_this)), _this.animatedStyles);
    };

    _this.updateValue = function (value) {
      return _this.flush(_this.value = value);
    };

    _this.interpolate = function (config, arg) {
      return new AnimatedInterpolation(_assertThisInitialized(_assertThisInitialized(_this)), config, arg);
    };

    _this.value = _value;
    _this.animatedStyles = new Set();
    _this.done = false;
    _this.startPosition = _value;
    _this.lastPosition = _value;
    _this.lastVelocity = undefined;
    _this.lastTime = undefined;
    _this.controller = undefined;
    return _this;
  }

  var _proto = AnimatedValue.prototype;

  _proto.flush = function flush() {
    if (this.animatedStyles.size === 0) this.updateStyles();
    this.animatedStyles.forEach(function (animatedStyle) {
      return animatedStyle.update();
    });
  };

  _proto.prepare = function prepare(controller) {
    // Values stay loyal to their original controller, this is also a way to
    // detect trailing values originating from a foreign controller
    if (this.controller === undefined) this.controller = controller;

    if (this.controller === controller) {
      this.startPosition = this.value;
      this.lastPosition = this.value;
      this.lastVelocity = controller.isActive ? this.lastVelocity : undefined;
      this.lastTime = controller.isActive ? this.lastTime : undefined;
      this.done = false;
      this.animatedStyles.clear();
    }
  };

  return AnimatedValue;
}(AnimatedWithChildren);

var AnimatedArray =
/*#__PURE__*/
function (_AnimatedArrayWithChi) {
  _inheritsLoose(AnimatedArray, _AnimatedArrayWithChi);

  function AnimatedArray(array) {
    var _this;

    _this = _AnimatedArrayWithChi.call(this) || this;

    _this.setValue = function (values) {
      return values.length === _this.payload.length && values.forEach(function (v, i) {
        return _this.payload[i].setValue(v);
      });
    };

    _this.getValue = function () {
      return _this.payload.map(function (v) {
        return v.getValue();
      });
    };

    _this.interpolate = function (config, arg) {
      return new AnimatedInterpolation(_assertThisInitialized(_assertThisInitialized(_this)), config, arg);
    };

    _this.payload = array instanceof AnimatedArray ? array.payload : array.map(function (n) {
      return new AnimatedValue(n);
    });
    return _this;
  }

  return AnimatedArray;
}(AnimatedArrayWithChildren);

function withDefault(value, defaultValue) {
  return value === undefined || value === null ? defaultValue : value;
}
function toArray(a) {
  return a !== void 0 ? Array.isArray(a) ? a : [a] : [];
}
function shallowEqual(a, b) {
  if (typeof a !== typeof b) return false;
  if (typeof a === 'string' || typeof a === 'number') return a === b;
  var i;

  for (i in a) {
    if (!(i in b)) return false;
  }

  for (i in b) {
    if (a[i] !== b[i]) return false;
  }

  return i === void 0 ? a === b : true;
}
function callProp(obj, state) {
  for (var _len = arguments.length, args = new Array(_len > 2 ? _len - 2 : 0), _key = 2; _key < _len; _key++) {
    args[_key - 2] = arguments[_key];
  }

  return typeof obj === 'function' ? obj.apply(void 0, [state].concat(args)) : obj;
}
function getValues$1(object) {
  return Object.keys(object).map(function (k) {
    return object[k];
  });
}
function getForwardProps(props) {
  var to = props.to,
      from = props.from,
      config = props.config,
      native = props.native,
      onStart = props.onStart,
      onRest = props.onRest,
      onFrame = props.onFrame,
      children = props.children,
      reset = props.reset,
      reverse = props.reverse,
      force = props.force,
      immediate = props.immediate,
      impl = props.impl,
      inject = props.inject,
      delay = props.delay,
      attach = props.attach,
      destroyed = props.destroyed,
      track = props.track,
      interpolateTo = props.interpolateTo,
      autoStart = props.autoStart,
      forward = _objectWithoutPropertiesLoose(props, ["to", "from", "config", "native", "onStart", "onRest", "onFrame", "children", "reset", "reverse", "force", "immediate", "impl", "inject", "delay", "attach", "destroyed", "track", "interpolateTo", "autoStart"]);

  return forward;
}
function interpolateTo(props) {
  var forward = getForwardProps(props);
  var rest = Object.keys(props).reduce(function (a, k) {
    var _extends2;

    return forward[k] !== void 0 ? a : _extends({}, a, (_extends2 = {}, _extends2[k] = props[k], _extends2));
  }, {});
  return _extends({
    to: forward
  }, rest);
}
function convertToAnimatedValue(acc, _ref) {
  var _extends3;

  var name = _ref[0],
      value = _ref[1];
  return _extends({}, acc, (_extends3 = {}, _extends3[name] = new (Array.isArray(value) ? AnimatedArray : AnimatedValue)(value), _extends3));
}
function convertValues(props) {
  var from = props.from,
      to = props.to,
      native = props.native;
  var allProps = Object.entries(_extends({}, from, to));
  return native ? allProps.reduce(convertToAnimatedValue, {}) : _extends({}, from, to);
}
function handleRef(ref, forward) {
  if (forward) {
    // If it's a function, assume it's a ref callback
    if (typeof forward === 'function') forward(ref);else if (typeof forward === 'object') // If it's an object and has a 'current' property, assume it's a ref object
      forward.current = ref;
  }

  return ref;
}

var AnimatedStyle =
/*#__PURE__*/
function (_AnimatedObjectWithCh) {
  _inheritsLoose(AnimatedStyle, _AnimatedObjectWithCh);

  function AnimatedStyle(style) {
    var _this;

    _this = _AnimatedObjectWithCh.call(this) || this;
    style = style || {};
    if (style.transform && !(style.transform instanceof Animated)) style = applyAnimatedValues.transform(style);
    _this.payload = style;
    return _this;
  }

  return AnimatedStyle;
}(AnimatedObjectWithChildren);

// http://www.w3.org/TR/css3-color/#svg-color
var colors = {
  transparent: 0x00000000,
  aliceblue: 0xf0f8ffff,
  antiquewhite: 0xfaebd7ff,
  aqua: 0x00ffffff,
  aquamarine: 0x7fffd4ff,
  azure: 0xf0ffffff,
  beige: 0xf5f5dcff,
  bisque: 0xffe4c4ff,
  black: 0x000000ff,
  blanchedalmond: 0xffebcdff,
  blue: 0x0000ffff,
  blueviolet: 0x8a2be2ff,
  brown: 0xa52a2aff,
  burlywood: 0xdeb887ff,
  burntsienna: 0xea7e5dff,
  cadetblue: 0x5f9ea0ff,
  chartreuse: 0x7fff00ff,
  chocolate: 0xd2691eff,
  coral: 0xff7f50ff,
  cornflowerblue: 0x6495edff,
  cornsilk: 0xfff8dcff,
  crimson: 0xdc143cff,
  cyan: 0x00ffffff,
  darkblue: 0x00008bff,
  darkcyan: 0x008b8bff,
  darkgoldenrod: 0xb8860bff,
  darkgray: 0xa9a9a9ff,
  darkgreen: 0x006400ff,
  darkgrey: 0xa9a9a9ff,
  darkkhaki: 0xbdb76bff,
  darkmagenta: 0x8b008bff,
  darkolivegreen: 0x556b2fff,
  darkorange: 0xff8c00ff,
  darkorchid: 0x9932ccff,
  darkred: 0x8b0000ff,
  darksalmon: 0xe9967aff,
  darkseagreen: 0x8fbc8fff,
  darkslateblue: 0x483d8bff,
  darkslategray: 0x2f4f4fff,
  darkslategrey: 0x2f4f4fff,
  darkturquoise: 0x00ced1ff,
  darkviolet: 0x9400d3ff,
  deeppink: 0xff1493ff,
  deepskyblue: 0x00bfffff,
  dimgray: 0x696969ff,
  dimgrey: 0x696969ff,
  dodgerblue: 0x1e90ffff,
  firebrick: 0xb22222ff,
  floralwhite: 0xfffaf0ff,
  forestgreen: 0x228b22ff,
  fuchsia: 0xff00ffff,
  gainsboro: 0xdcdcdcff,
  ghostwhite: 0xf8f8ffff,
  gold: 0xffd700ff,
  goldenrod: 0xdaa520ff,
  gray: 0x808080ff,
  green: 0x008000ff,
  greenyellow: 0xadff2fff,
  grey: 0x808080ff,
  honeydew: 0xf0fff0ff,
  hotpink: 0xff69b4ff,
  indianred: 0xcd5c5cff,
  indigo: 0x4b0082ff,
  ivory: 0xfffff0ff,
  khaki: 0xf0e68cff,
  lavender: 0xe6e6faff,
  lavenderblush: 0xfff0f5ff,
  lawngreen: 0x7cfc00ff,
  lemonchiffon: 0xfffacdff,
  lightblue: 0xadd8e6ff,
  lightcoral: 0xf08080ff,
  lightcyan: 0xe0ffffff,
  lightgoldenrodyellow: 0xfafad2ff,
  lightgray: 0xd3d3d3ff,
  lightgreen: 0x90ee90ff,
  lightgrey: 0xd3d3d3ff,
  lightpink: 0xffb6c1ff,
  lightsalmon: 0xffa07aff,
  lightseagreen: 0x20b2aaff,
  lightskyblue: 0x87cefaff,
  lightslategray: 0x778899ff,
  lightslategrey: 0x778899ff,
  lightsteelblue: 0xb0c4deff,
  lightyellow: 0xffffe0ff,
  lime: 0x00ff00ff,
  limegreen: 0x32cd32ff,
  linen: 0xfaf0e6ff,
  magenta: 0xff00ffff,
  maroon: 0x800000ff,
  mediumaquamarine: 0x66cdaaff,
  mediumblue: 0x0000cdff,
  mediumorchid: 0xba55d3ff,
  mediumpurple: 0x9370dbff,
  mediumseagreen: 0x3cb371ff,
  mediumslateblue: 0x7b68eeff,
  mediumspringgreen: 0x00fa9aff,
  mediumturquoise: 0x48d1ccff,
  mediumvioletred: 0xc71585ff,
  midnightblue: 0x191970ff,
  mintcream: 0xf5fffaff,
  mistyrose: 0xffe4e1ff,
  moccasin: 0xffe4b5ff,
  navajowhite: 0xffdeadff,
  navy: 0x000080ff,
  oldlace: 0xfdf5e6ff,
  olive: 0x808000ff,
  olivedrab: 0x6b8e23ff,
  orange: 0xffa500ff,
  orangered: 0xff4500ff,
  orchid: 0xda70d6ff,
  palegoldenrod: 0xeee8aaff,
  palegreen: 0x98fb98ff,
  paleturquoise: 0xafeeeeff,
  palevioletred: 0xdb7093ff,
  papayawhip: 0xffefd5ff,
  peachpuff: 0xffdab9ff,
  peru: 0xcd853fff,
  pink: 0xffc0cbff,
  plum: 0xdda0ddff,
  powderblue: 0xb0e0e6ff,
  purple: 0x800080ff,
  rebeccapurple: 0x663399ff,
  red: 0xff0000ff,
  rosybrown: 0xbc8f8fff,
  royalblue: 0x4169e1ff,
  saddlebrown: 0x8b4513ff,
  salmon: 0xfa8072ff,
  sandybrown: 0xf4a460ff,
  seagreen: 0x2e8b57ff,
  seashell: 0xfff5eeff,
  sienna: 0xa0522dff,
  silver: 0xc0c0c0ff,
  skyblue: 0x87ceebff,
  slateblue: 0x6a5acdff,
  slategray: 0x708090ff,
  slategrey: 0x708090ff,
  snow: 0xfffafaff,
  springgreen: 0x00ff7fff,
  steelblue: 0x4682b4ff,
  tan: 0xd2b48cff,
  teal: 0x008080ff,
  thistle: 0xd8bfd8ff,
  tomato: 0xff6347ff,
  turquoise: 0x40e0d0ff,
  violet: 0xee82eeff,
  wheat: 0xf5deb3ff,
  white: 0xffffffff,
  whitesmoke: 0xf5f5f5ff,
  yellow: 0xffff00ff,
  yellowgreen: 0x9acd32ff
};

// const INTEGER = '[-+]?\\d+';
var NUMBER = '[-+]?\\d*\\.?\\d+';
var PERCENTAGE = NUMBER + '%';

function call() {
  return '\\(\\s*(' + Array.prototype.slice.call(arguments).join(')\\s*,\\s*(') + ')\\s*\\)';
}

var rgb = new RegExp('rgb' + call(NUMBER, NUMBER, NUMBER));
var rgba = new RegExp('rgba' + call(NUMBER, NUMBER, NUMBER, NUMBER));
var hsl = new RegExp('hsl' + call(NUMBER, PERCENTAGE, PERCENTAGE));
var hsla = new RegExp('hsla' + call(NUMBER, PERCENTAGE, PERCENTAGE, NUMBER));
var hex3 = /^#([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})$/;
var hex4 = /^#([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})$/;
var hex6 = /^#([0-9a-fA-F]{6})$/;
var hex8 = /^#([0-9a-fA-F]{8})$/;

/*
https://github.com/react-community/normalize-css-color

BSD 3-Clause License

Copyright (c) 2016, React Community
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of the copyright holder nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
function normalizeColor(color) {
  var match;

  if (typeof color === 'number') {
    return color >>> 0 === color && color >= 0 && color <= 0xffffffff ? color : null;
  } // Ordered based on occurrences on Facebook codebase


  if (match = hex6.exec(color)) return parseInt(match[1] + 'ff', 16) >>> 0;
  if (colors.hasOwnProperty(color)) return colors[color];

  if (match = rgb.exec(color)) {
    return (parse255(match[1]) << 24 | // r
    parse255(match[2]) << 16 | // g
    parse255(match[3]) << 8 | // b
    0x000000ff) >>> // a
    0;
  }

  if (match = rgba.exec(color)) {
    return (parse255(match[1]) << 24 | // r
    parse255(match[2]) << 16 | // g
    parse255(match[3]) << 8 | // b
    parse1(match[4])) >>> // a
    0;
  }

  if (match = hex3.exec(color)) {
    return parseInt(match[1] + match[1] + // r
    match[2] + match[2] + // g
    match[3] + match[3] + // b
    'ff', // a
    16) >>> 0;
  } // https://drafts.csswg.org/css-color-4/#hex-notation


  if (match = hex8.exec(color)) return parseInt(match[1], 16) >>> 0;

  if (match = hex4.exec(color)) {
    return parseInt(match[1] + match[1] + // r
    match[2] + match[2] + // g
    match[3] + match[3] + // b
    match[4] + match[4], // a
    16) >>> 0;
  }

  if (match = hsl.exec(color)) {
    return (hslToRgb(parse360(match[1]), // h
    parsePercentage(match[2]), // s
    parsePercentage(match[3]) // l
    ) | 0x000000ff) >>> // a
    0;
  }

  if (match = hsla.exec(color)) {
    return (hslToRgb(parse360(match[1]), // h
    parsePercentage(match[2]), // s
    parsePercentage(match[3]) // l
    ) | parse1(match[4])) >>> // a
    0;
  }

  return null;
}

function hue2rgb(p, q, t) {
  if (t < 0) t += 1;
  if (t > 1) t -= 1;
  if (t < 1 / 6) return p + (q - p) * 6 * t;
  if (t < 1 / 2) return q;
  if (t < 2 / 3) return p + (q - p) * (2 / 3 - t) * 6;
  return p;
}

function hslToRgb(h, s, l) {
  var q = l < 0.5 ? l * (1 + s) : l + s - l * s;
  var p = 2 * l - q;
  var r = hue2rgb(p, q, h + 1 / 3);
  var g = hue2rgb(p, q, h);
  var b = hue2rgb(p, q, h - 1 / 3);
  return Math.round(r * 255) << 24 | Math.round(g * 255) << 16 | Math.round(b * 255) << 8;
}

function parse255(str) {
  var int = parseInt(str, 10);
  if (int < 0) return 0;
  if (int > 255) return 255;
  return int;
}

function parse360(str) {
  var int = parseFloat(str);
  return (int % 360 + 360) % 360 / 360;
}

function parse1(str) {
  var num = parseFloat(str);
  if (num < 0) return 0;
  if (num > 1) return 255;
  return Math.round(num * 255);
}

function parsePercentage(str) {
  // parseFloat conveniently ignores the final %
  var int = parseFloat(str);
  if (int < 0) return 0;
  if (int > 100) return 1;
  return int / 100;
}

function colorToRgba(input) {
  var int32Color = normalizeColor(input);
  if (int32Color === null) return input;
  int32Color = int32Color || 0;
  var r = (int32Color & 0xff000000) >>> 24;
  var g = (int32Color & 0x00ff0000) >>> 16;
  var b = (int32Color & 0x0000ff00) >>> 8;
  var a = (int32Color & 0x000000ff) / 255;
  return "rgba(" + r + ", " + g + ", " + b + ", " + a + ")";
} // Problem: https://github.com/animatedjs/animated/pull/102
// Solution: https://stackoverflow.com/questions/638565/parsing-scientific-notation-sensibly/658662


var stringShapeRegex = /[+\-]?(?:0|[1-9]\d*)(?:\.\d*)?(?:[eE][+\-]?\d+)?/g; // Covers rgb, rgba, hsl, hsla
// Taken from https://gist.github.com/olmokramer/82ccce673f86db7cda5e

var colorRegex = /(#(?:[0-9a-f]{2}){2,4}|(#[0-9a-f]{3})|(rgb|hsl)a?\((-?\d+%?[,\s]+){2,3}\s*[\d\.]+%?\))/gi; // Covers color names (transparent, blue, etc.)

var colorNamesRegex = new RegExp("(" + Object.keys(colors).join('|') + ")", 'g');
/**
 * Supports string shapes by extracting numbers so new values can be computed,
 * and recombines those values into new strings of the same shape.  Supports
 * things like:
 *
 *   rgba(123, 42, 99, 0.36)           // colors
 *   -45deg                            // values with units
 *   0 2px 2px 0px rgba(0, 0, 0, 0.12) // box shadows
 */

function createInterpolation(config) {
  // Replace colors with rgba
  var outputRange = config.output.map(function (rangeValue) {
    return rangeValue.replace(colorRegex, colorToRgba);
  }).map(function (rangeValue) {
    return rangeValue.replace(colorNamesRegex, colorToRgba);
  }); // ->
  // [
  //   [0, 50],
  //   [100, 150],
  //   [200, 250],
  //   [0, 0.5],
  // ]

  var outputRanges = outputRange[0].match(stringShapeRegex).map(function () {
    return [];
  });
  outputRange.forEach(function (value) {
    value.match(stringShapeRegex).forEach(function (number, i) {
      return outputRanges[i].push(+number);
    });
  });
  var interpolations = outputRange[0].match(stringShapeRegex).map(function (value, i) {
    return Interpolation.create(_extends({}, config, {
      output: outputRanges[i]
    }));
  });
  var shouldRound = /^rgb/.test(outputRange[0]);
  return function (input) {
    var i = 0;
    return outputRange[0] // 'rgba(0, 100, 200, 0)'
    // ->
    // 'rgba(${interpolations[0](input)}, ${interpolations[1](input)}, ...'
    .replace(stringShapeRegex, function () {
      return interpolations[i++](input);
    }) // rgba requires that the r,g,b are integers.... so we want to round them, but we *dont* want to
    // round the opacity (4th column).
    .replace(/rgba\(([0-9\.-]+), ([0-9\.-]+), ([0-9\.-]+), ([0-9\.-]+)\)/gi, function (_, p1, p2, p3, p4) {
      return "rgba(" + Math.round(p1) + ", " + Math.round(p2) + ", " + Math.round(p3) + ", " + p4 + ")";
    });
  };
}

var check = function check(value) {
  return value === 'auto';
};

var overwrite = function overwrite(width, height) {
  return function (acc, _ref) {
    var _extends2;

    var name = _ref[0],
        value = _ref[1];
    return _extends({}, acc, (_extends2 = {}, _extends2[name] = value === 'auto' ? ~name.indexOf('height') ? height : width : value, _extends2));
  };
};

function fixAuto(props, callback) {
  var from = props.from,
      to = props.to,
      children = props.children; // Dry-route props back if nothing's using 'auto' in there
  // TODO: deal with "null"

  if (!(getValues$1(to).some(check) || getValues$1(from).some(check))) return; // Fetch render v-dom

  var element = children(convertValues(props)); // A spring can return undefined/null, check against that (#153)

  if (!element) return;
  var elementStyles = element.props.style; // Return v.dom with injected ref

  return React.createElement(element.type, _extends({
    key: element.key ? element.key : undefined
  }, element.props, {
    style: _extends({}, elementStyles, {
      position: 'absolute',
      visibility: 'hidden'
    }),
    ref: function ref(_ref2) {
      if (_ref2) {
        // Once it's rendered out, fetch bounds (minus padding/margin/borders)
        var node = ReactDOM.findDOMNode(_ref2);
        var width, height;
        var cs = getComputedStyle(node);

        if (cs.boxSizing === 'border-box') {
          width = node.offsetWidth;
          height = node.offsetHeight;
        } else {
          var paddingX = parseFloat(cs.paddingLeft || 0) + parseFloat(cs.paddingRight || 0);
          var paddingY = parseFloat(cs.paddingTop || 0) + parseFloat(cs.paddingBottom || 0);
          var borderX = parseFloat(cs.borderLeftWidth || 0) + parseFloat(cs.borderRightWidth || 0);
          var borderY = parseFloat(cs.borderTopWidth || 0) + parseFloat(cs.borderBottomWidth || 0);
          width = node.offsetWidth - paddingX - borderX;
          height = node.offsetHeight - paddingY - borderY;
        }

        var convert = overwrite(width, height);
        callback(_extends({}, props, {
          from: Object.entries(from).reduce(convert, from),
          to: Object.entries(to).reduce(convert, to)
        }));
      }
    }
  }));
}

var isUnitlessNumber = {
  animationIterationCount: true,
  borderImageOutset: true,
  borderImageSlice: true,
  borderImageWidth: true,
  boxFlex: true,
  boxFlexGroup: true,
  boxOrdinalGroup: true,
  columnCount: true,
  columns: true,
  flex: true,
  flexGrow: true,
  flexPositive: true,
  flexShrink: true,
  flexNegative: true,
  flexOrder: true,
  gridRow: true,
  gridRowEnd: true,
  gridRowSpan: true,
  gridRowStart: true,
  gridColumn: true,
  gridColumnEnd: true,
  gridColumnSpan: true,
  gridColumnStart: true,
  fontWeight: true,
  lineClamp: true,
  lineHeight: true,
  opacity: true,
  order: true,
  orphans: true,
  tabSize: true,
  widows: true,
  zIndex: true,
  zoom: true,
  // SVG-related properties
  fillOpacity: true,
  floodOpacity: true,
  stopOpacity: true,
  strokeDasharray: true,
  strokeDashoffset: true,
  strokeMiterlimit: true,
  strokeOpacity: true,
  strokeWidth: true
};

var prefixKey = function prefixKey(prefix, key) {
  return prefix + key.charAt(0).toUpperCase() + key.substring(1);
};

var prefixes = ['Webkit', 'Ms', 'Moz', 'O'];
isUnitlessNumber = Object.keys(isUnitlessNumber).reduce(function (acc, prop) {
  prefixes.forEach(function (prefix) {
    return acc[prefixKey(prefix, prop)] = acc[prop];
  });
  return acc;
}, isUnitlessNumber);

function dangerousStyleValue(name, value, isCustomProperty) {
  if (value == null || typeof value === 'boolean' || value === '') return '';
  if (!isCustomProperty && typeof value === 'number' && value !== 0 && !(isUnitlessNumber.hasOwnProperty(name) && isUnitlessNumber[name])) return value + 'px'; // Presumes implicit 'px' suffix for unitless numbers

  return ('' + value).trim();
}

var attributeCache = {};
injectCreateAnimatedStyle(function (style) {
  return new AnimatedStyle(style);
});
injectDefaultElement('div');
injectInterpolation(createInterpolation);
injectColorNames(colors);
injectBugfixes(fixAuto);
injectApplyAnimatedValues(function (instance, props) {
  if (instance.nodeType && instance.setAttribute !== undefined) {
    var style = props.style,
        children = props.children,
        scrollTop = props.scrollTop,
        scrollLeft = props.scrollLeft,
        attributes = _objectWithoutPropertiesLoose(props, ["style", "children", "scrollTop", "scrollLeft"]);

    if (scrollTop !== void 0) instance.scrollTop = scrollTop;
    if (scrollLeft !== void 0) instance.scrollLeft = scrollLeft; // Set textContent, if children is an animatable value

    if (children !== void 0) instance.textContent = children; // Set styles ...

    for (var styleName in style) {
      if (!style.hasOwnProperty(styleName)) continue;
      var isCustomProperty = styleName.indexOf('--') === 0;
      var styleValue = dangerousStyleValue(styleName, style[styleName], isCustomProperty);
      if (styleName === 'float') styleName = 'cssFloat';
      if (isCustomProperty) instance.style.setProperty(styleName, styleValue);else instance.style[styleName] = styleValue;
    } // Set attributes ...


    for (var name in attributes) {
      // Attributes are written in dash case
      var dashCase = attributeCache[name] || (attributeCache[name] = name.replace(/([A-Z])/g, function (n) {
        return '-' + n.toLowerCase();
      }));
      if (typeof instance.getAttribute(dashCase) !== 'undefined') instance.setAttribute(dashCase, attributes[name]);
    }
  } else return false;
}, function (style) {
  return style;
});

var AnimatedProps =
/*#__PURE__*/
function (_AnimatedObjectWithCh) {
  _inheritsLoose(AnimatedProps, _AnimatedObjectWithCh);

  function AnimatedProps(props, callback) {
    var _this;

    _this = _AnimatedObjectWithCh.call(this) || this;
    if (props.style) props = _extends({}, props, {
      style: createAnimatedStyle(props.style)
    });
    _this.payload = props;
    _this.update = callback;

    _this.attach();

    return _this;
  }

  return AnimatedProps;
}(AnimatedObjectWithChildren);

function createAnimatedComponent(Component) {
  var AnimatedComponent =
  /*#__PURE__*/
  function (_React$Component) {
    _inheritsLoose(AnimatedComponent, _React$Component);

    function AnimatedComponent(props) {
      var _this;

      _this = _React$Component.call(this) || this;

      _this.callback = function () {
        if (_this.node) {
          var didUpdate = applyAnimatedValues.fn(_this.node, _this.propsAnimated.getAnimatedValue(), _assertThisInitialized(_assertThisInitialized(_this)));
          if (didUpdate === false) _this.forceUpdate();
        }
      };

      _this.attachProps(props);

      return _this;
    }

    var _proto = AnimatedComponent.prototype;

    _proto.componentWillUnmount = function componentWillUnmount() {
      this.propsAnimated && this.propsAnimated.detach();
    };

    _proto.setNativeProps = function setNativeProps(props) {
      var didUpdate = applyAnimatedValues.fn(this.node, props, this);
      if (didUpdate === false) this.forceUpdate();
    }; // The system is best designed when setNativeProps is implemented. It is
    // able to avoid re-rendering and directly set the attributes that
    // changed. However, setNativeProps can only be implemented on leaf
    // native components. If you want to animate a composite component, you
    // need to re-render it. In this case, we have a fallback that uses
    // forceUpdate.


    _proto.attachProps = function attachProps(_ref) {
      var forwardRef = _ref.forwardRef,
          nextProps = _objectWithoutPropertiesLoose(_ref, ["forwardRef"]);

      var oldPropsAnimated = this.propsAnimated;
      this.propsAnimated = new AnimatedProps(nextProps, this.callback); // When you call detach, it removes the element from the parent list
      // of children. If it goes to 0, then the parent also detaches itself
      // and so on.
      // An optimization is to attach the new elements and THEN detach the old
      // ones instead of detaching and THEN attaching.
      // This way the intermediate state isn't to go to 0 and trigger
      // this expensive recursive detaching to then re-attach everything on
      // the very next operation.

      oldPropsAnimated && oldPropsAnimated.detach();
    };

    _proto.shouldComponentUpdate = function shouldComponentUpdate(props) {
      var style = props.style,
          nextProps = _objectWithoutPropertiesLoose(props, ["style"]);

      var _this$props = this.props,
          currentStyle = _this$props.style,
          currentProps = _objectWithoutPropertiesLoose(_this$props, ["style"]);

      if (!shallowEqual(currentProps, nextProps) || !shallowEqual(currentStyle, style)) {
        this.attachProps(props);
        return true;
      }

      return false;
    };

    _proto.render = function render() {
      var _this2 = this;

      var forwardRef = this.props.forwardRef;

      var _this$propsAnimated$g = this.propsAnimated.getValue(),
          scrollTop = _this$propsAnimated$g.scrollTop,
          scrollLeft = _this$propsAnimated$g.scrollLeft,
          animatedProps = _objectWithoutPropertiesLoose(_this$propsAnimated$g, ["scrollTop", "scrollLeft"]);

      return React.createElement(Component, _extends({}, animatedProps, {
        ref: function ref(node) {
          return _this2.node = handleRef(node, _this2.props.forwardRef);
        }
      }));
    };

    return AnimatedComponent;
  }(React.Component);

  return React.forwardRef(function (props, ref) {
    return React.createElement(AnimatedComponent, _extends({}, props, {
      forwardRef: ref
    }));
  });
}

var config = {
  default: {
    tension: 170,
    friction: 26
  },
  gentle: {
    tension: 120,
    friction: 14
  },
  wobbly: {
    tension: 180,
    friction: 12
  },
  stiff: {
    tension: 210,
    friction: 20
  },
  slow: {
    tension: 280,
    friction: 60
  },
  molasses: {
    tension: 280,
    friction: 120
  }
};

var now$1, isDone, noChange, configIdx, valIdx, config$1, animation, position, from, tracked, to, endOfAnimation, lastTime, velocity, numSteps, force, damping, acceleration, stepIdx, isOvershooting, isVelocity, isDisplacement;

var Controller =
/*#__PURE__*/
function () {
  function Controller(props, _config) {
    var _this = this;

    if (_config === void 0) {
      _config = {
        native: true,
        track: true,
        interpolateTo: true,
        autoStart: true
      };
    }

    this.getValues = function () {
      return _this.props.native ? _this.interpolations : _this.animatedProps;
    };

    this.raf = function () {
      now$1 = now();
      isDone = true;
      noChange = true;

      for (configIdx = 0; configIdx < _this.configs.length; configIdx++) {
        config$1 = _this.configs[configIdx]; // Doing delay here instead of setTimeout is one async worry less

        if (config$1.delay && now$1 - _this.startTime < config$1.delay) {
          isDone = false;
          continue;
        }

        for (valIdx = 0; valIdx < config$1.animatedValues.length; valIdx++) {
          animation = config$1.animatedValues[valIdx];
          position = animation.lastPosition;
          from = config$1.fromValues[valIdx];
          tracked = config$1.parent.track && config$1.parent.track.getPayload(valIdx);
          to = tracked ? tracked.getValue() : config$1.toValues[valIdx]; // If an animation is done, skip, until all of them conclude

          if (animation.done) continue; // Break animation when animation is immediate or string values are involved

          if (config$1.immediate || typeof from === 'string' || typeof to === 'string') {
            animation.updateValue(to);
            animation.done = true;
            continue;
          } else noChange = false;

          if (config$1.duration !== void 0) {
            position = from + config$1.easing((now$1 - _this.startTime - config$1.delay) / config$1.duration) * (to - from);
            endOfAnimation = now$1 >= _this.startTime + config$1.delay + config$1.duration;
          } else {
            lastTime = animation.lastTime !== void 0 ? animation.lastTime : now$1;
            velocity = animation.lastVelocity !== void 0 ? animation.lastVelocity : config$1.initialVelocity; // If we lost a lot of frames just jump to the end.

            if (now$1 > lastTime + 64) lastTime = now$1; // http://gafferongames.com/game-physics/fix-your-timestep/

            numSteps = Math.floor(now$1 - lastTime);

            for (stepIdx = 0; stepIdx < numSteps; ++stepIdx) {
              force = -config$1.tension * (position - to);
              damping = -config$1.friction * velocity;
              acceleration = (force + damping) / config$1.mass;
              velocity = velocity + acceleration * 1 / 1000;
              position = position + velocity * 1 / 1000;
            } // Conditions for stopping the spring animation


            isOvershooting = config$1.clamp && config$1.tension !== 0 ? from < to ? position > to : position < to : false;
            isVelocity = Math.abs(velocity) <= config$1.precision;
            isDisplacement = config$1.tension !== 0 ? Math.abs(to - position) <= config$1.precision : true;
            endOfAnimation = isOvershooting || isVelocity && isDisplacement;
            animation.lastVelocity = velocity;
            animation.lastTime = now$1;
          } // Trails aren't done until their parents conclude


          if (config$1.parent.track && !tracked.done) endOfAnimation = false;

          if (endOfAnimation) {
            // Ensure that we end up with a round value
            if (animation.value !== to) position = to;
            animation.done = true;
          } else isDone = false;

          animation.updateValue(position);
          animation.lastPosition = position;
        } // Keep track of updated values only when necessary


        if (_this.props.onFrame || !_this.props.native) _this.animatedProps[config$1.name] = config$1.interpolation.getValue();
      } // Update callbacks in the end of the frame


      if (_this.props.onFrame || !_this.props.native) {
        if (!_this.props.native && _this.onUpdate) _this.onUpdate();
        if (_this.props.onFrame) _this.props.onFrame(_this.animatedProps);
      } // Either call onEnd or next frame


      if (isDone) return _this.debouncedOnEnd({
        finished: true,
        noChange: noChange
      });
      _this.frame = requestFrame(_this.raf);
    };

    this.dependents = new Set();
    this.isActive = false;
    this.hasChanged = false;
    this.props = {};
    this.merged = {};
    this.animations = {};
    this.interpolations = {};
    this.animatedProps = {};
    this.configs = [];
    this.frame = undefined;
    this.startTime = undefined;
    this.lastTime = undefined;
    this.update(_extends({}, props, _config));
  }
  /**
   * props: to: { ... }
   *
   *  { name: value }             pairs
   *  { name: AnimatedValue() }   animated values
   *  { name: [1,2,3] }           plain numeric arrays
   *  { name: AnimatedArray() }   animated arrays
   *
   * Plain values can be:
   *
   *  123                         Numbers
   *  "hello"                     Strings
   *  "#3d4d5d" ...               Colors (rga, rgba, hex, plain names)
   *  "something 12 something"    Interpolation patterns with numbers in them
   *
   * Additionally, springs are allowed to "attach" to another AnimationController, fetching the
   * values from there, if present.
   */


  var _proto = Controller.prototype;

  _proto.update = function update(props) {
    var _this2 = this;

    this.props = _extends({}, this.props, props);

    var _ref = this.props.interpolateTo ? interpolateTo(this.props) : this.props,
        _ref$from = _ref.from,
        from = _ref$from === void 0 ? {} : _ref$from,
        _ref$to = _ref.to,
        to = _ref$to === void 0 ? {} : _ref$to,
        _ref$config = _ref.config,
        config = _ref$config === void 0 ? {} : _ref$config,
        _ref$delay = _ref.delay,
        delay = _ref$delay === void 0 ? 0 : _ref$delay,
        reverse = _ref.reverse,
        attach = _ref.attach,
        reset = _ref.reset,
        immediate = _ref.immediate,
        native = _ref.native,
        onFrame = _ref.onFrame,
        track = _ref.track,
        autoStart = _ref.autoStart; // Reverse values when requested


    if (reverse) {
      var _ref2 = [to, from];
      from = _ref2[0];
      to = _ref2[1];
    }

    this.hasChanged = false; // Attachment handling, trailed springs can "attach" themselves to a previous spring

    var target = attach && attach(this); // Reset merged props when necessary

    var extra = reset ? {} : this.merged; // This will collect all props that were ever set

    this.merged = _extends({}, from, extra, to); // Reduces input { name: value } pairs into animated values

    this.animations = Object.entries(this.merged).reduce(function (acc, _ref3, i) {
      var name = _ref3[0],
          value = _ref3[1];
      // Issue cached entries, except on reset
      var entry = !reset && acc[name] || {}; // Attach allows a spring to fetch its values elsewhere

      var _value = value,
          _entry;

      if (target && target.animations[name]) {
        _entry = target.animations[name];
        value = _entry.parent;
      } // Figure out what the value is supposed to be


      var isArray, isString, isNumber, isInterpolation;
      var isAnimated = value instanceof Animated;

      if (!isAnimated) {
        isArray = Array.isArray(value);
        isNumber = typeof value === 'number';
        isString = typeof value === 'string' && !value.startsWith('#') && !/\d/.test(value) && !colorNames[value];
        isInterpolation = !isNumber && !isString && !isArray;
      } // Carry actual values (including animated) in order to change detect


      var changes = isAnimated ? value.getPayload() : value; // Detect changes, animated values will be checked in the raf-loop

      if (isAnimated || !shallowEqual(entry.changes, changes)) {
        var _extends2;

        _this2.hasChanged = true;
        var parent, interpolation$$1;

        var _from = from[name] !== void 0 ? from[name] : value;

        var _config2 = callProp(config, name);

        if (isAnimated) {
          // We end up here if the value we're shifting to is an animated value or array
          parent = entry.parent || new value.constructor(value.getValue());
          interpolation$$1 = entry.interpolation || parent; // In the next step we're going to check if that value is interpolated

          if (_entry && _entry.interpolation.calc) {
            var _config3 = {
              output: [interpolation$$1.calc ? interpolation$$1.calc(parent.value) : _entry.interpolation.calc(0), _value]
            };
            if (interpolation$$1.calc) interpolation$$1.updateConfig(_config3);else interpolation$$1 = parent.interpolate(_config3);
            parent.value = 0;
          }
        } else if (isNumber || isString) {
          parent = interpolation$$1 = entry.parent || new AnimatedValue(_from);
        } else if (isArray) {
          parent = interpolation$$1 = entry.parent || new AnimatedArray(_from);
        } else if (isInterpolation) {
          // Deal with interpolations
          var prev = entry.interpolation && entry.interpolation.calc(entry.parent.value); // Interpolations are not addaptive, start with 0

          if (entry.parent) {
            parent = entry.parent;
            parent.value = 0;
          } else parent = new AnimatedValue(0); // Map from-to on a scale between 0-1


          var range = {
            output: [prev !== void 0 ? prev : _from, value]
          };

          if (entry.interpolation) {
            interpolation$$1 = entry.interpolation;
            entry.interpolation.updateConfig(range);
          } else interpolation$$1 = parent.interpolate(range); // And stop at 1


          value = 1;
        }

        parent.controller = _this2;
        if (isAnimated && value.controller !== _this2) value.controller.dependents.add(_this2);
        parent.track = isAnimated ? value : undefined; // Set immediate values

        if (callProp(immediate, name)) parent.value = value; // Map output values to an array so reading out is easier later on

        var animatedValues = toArray(parent.getPayload());
        var fromValues = toArray(parent.getValue());
        var toValues = toArray(isAnimated ? value.getValue() : value); // Reset animated values

        animatedValues.forEach(function (value) {
          return value.prepare(_this2);
        });
        return _extends({}, acc, (_extends2 = {}, _extends2[name] = _extends({}, entry, {
          name: name,
          parent: parent,
          // The animated object on which the update-cb is called
          interpolation: interpolation$$1,
          // The parents interpolation, if any. If not it refers to the parent
          animatedValues: animatedValues,
          // An array of all animated values taking part in this op
          fromValues: fromValues,
          // Raw/numerical start-state values
          toValues: toValues,
          // Raw/numerical/end-state values
          changes: changes,
          immediate: callProp(immediate, name),
          delay: withDefault(_config2.delay, delay || 0),
          initialVelocity: withDefault(_config2.velocity, 0),
          clamp: withDefault(_config2.clamp, false),
          precision: withDefault(_config2.precision, 0.01),
          tension: withDefault(_config2.tension, 170),
          friction: withDefault(_config2.friction, 26),
          mass: withDefault(_config2.mass, 1),
          duration: _config2.duration,
          easing: withDefault(_config2.easing, function (t) {
            return t;
          })
        }), _extends2));
      } else return acc;
    }, this.animations);

    if (this.hasChanged) {
      this.configs = getValues$1(this.animations);
      this.animatedProps = {};
      this.interpolations = {};

      for (var key in this.animations) {
        this.interpolations[key] = this.animations[key].interpolation;
        this.animatedProps[key] = this.animations[key].interpolation.getValue();
      }
    }

    for (var _len = arguments.length, start = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      start[_key - 1] = arguments[_key];
    }

    if (autoStart || start.length) this.start.apply(this, start);
    return this.getValues();
  };

  _proto.start = function start(onEnd, onUpdate) {
    this.startTime = now();
    if (this.isActive) this.stop();
    this.isActive = true;
    this.onEnd = typeof onEnd === 'function' && onEnd;
    this.onUpdate = onUpdate;
    if (this.props.onStart) this.props.onStart(); // Start RAF loop

    this.frame = requestFrame(this.raf); // Call dependent controllers

    if (this.props.track) {
      for (var _iterator = this.dependents, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
        var _ref4;

        if (_isArray) {
          if (_i >= _iterator.length) break;
          _ref4 = _iterator[_i++];
        } else {
          _i = _iterator.next();
          if (_i.done) break;
          _ref4 = _i.value;
        }

        var controller = _ref4;
        controller.update(_extends({}, controller.props, controller.merged), true);
      }
    }
  };

  _proto.stop = function stop(finished) {
    if (finished === void 0) {
      finished = false;
    }

    config$1 = undefined;
    animation = undefined;
    from = undefined;
    tracked = undefined;
    to = undefined; // Reset collected changes since the animation has been stopped cold turkey

    if (finished) getValues$1(this.animations).forEach(function (a) {
      return a.changes = undefined;
    });
    this.isActive = false;
    cancelFrame(this.frame);
    this.debouncedOnEnd({
      finished: finished
    });
  };

  _proto.debouncedOnEnd = function debouncedOnEnd(result) {
    this.isActive = false;
    var onEnd = this.onEnd;
    this.onEnd = null;
    onEnd && onEnd(result);
  };

  return Controller;
}();

function useSpring(_ref) {
  var onRest = _ref.onRest,
      _ref$onKeyframesHalt = _ref.onKeyframesHalt,
      onKeyframesHalt = _ref$onKeyframesHalt === void 0 ? function () {
    return null;
  } : _ref$onKeyframesHalt,
      _ref$updatePropsOnRer = _ref.updatePropsOnRerender,
      updatePropsOnRerender = _ref$updatePropsOnRer === void 0 ? true : _ref$updatePropsOnRer,
      props = _objectWithoutPropertiesLoose(_ref, ["onRest", "onKeyframesHalt", "updatePropsOnRerender"]);

  var _React$useState = React.useState(new Controller(props)),
      ctrl = _React$useState[0];

  var _React$useState2 = React.useState(),
      forceUpdate = _React$useState2[1];

  var onHalt = onRest ? function (_ref2) {
    var finished = _ref2.finished;
    return finished && onRest(ctrl.merged);
  } : onKeyframesHalt(ctrl);
  var update = React.useCallback( // resolve and last are passed to the update function from the keyframes controller
  function (animProps) {
    ctrl.update(animProps, onHalt);
    requestFrame(function () {
      return animProps.reset && forceUpdate();
    });
  }, [onRest, onKeyframesHalt]);
  React.useEffect(function () {
    return void (updatePropsOnRerender && update(props));
  });
  return [ctrl.getValues(), function (props) {
    return update(props);
  }, function (finished) {
    if (finished === void 0) {
      finished = false;
    }

    return ctrl.stop(finished);
  }];
}

function useTrail(_ref) {
  var items = _ref.items,
      delay = _ref.delay,
      reverse = _ref.reverse,
      _ref$onKeyframesHalt = _ref.onKeyframesHalt,
      onKeyframesHalt = _ref$onKeyframesHalt === void 0 ? function () {
    return null;
  } : _ref$onKeyframesHalt,
      _ref$updatePropsOnRer = _ref.updatePropsOnRerender,
      updatePropsOnRerender = _ref$updatePropsOnRer === void 0 ? true : _ref$updatePropsOnRer,
      onRest = _ref.onRest,
      props = _objectWithoutPropertiesLoose(_ref, ["items", "delay", "reverse", "onKeyframesHalt", "updatePropsOnRerender", "onRest"]);

  var instances = React.useRef();
  var prevItems = React.useRef();
  var array = toArray(items);
  var mounted = React.useRef(false);

  var _React$useState = React.useState(),
      forceUpdate = _React$useState[1];

  var onHalt = onRest ? function (ctrl) {
    return function (_ref2) {
      var finished = _ref2.finished;
      finished && mounted.current && onRest(ctrl.merged);
    };
  } : onKeyframesHalt;

  if (prevItems.current !== items) {
    instances.current = new Map(array.map(function (_, idx) {
      return [idx, new Controller(props)];
    }));
  }

  var update = React.useCallback(
  /** resolve and last are passed to the update function from the keyframes controller */
  function (props) {
    var _loop = function _loop() {
      if (_isArray) {
        if (_i >= _iterator.length) return "break";
        _ref3 = _iterator[_i++];
      } else {
        _i = _iterator.next();
        if (_i.done) return "break";
        _ref3 = _i.value;
      }

      var _ref4 = _ref3,
          idx = _ref4[0],
          ctrl = _ref4[1];

      if (idx === 0) {
        ctrl.update(_extends({
          delay: delay
        }, props, {
          attach: undefined
        }));
      }

      ctrl.update(_extends({}, props, {
        attach: function attach() {
          return instances.current.get(idx - 1);
        }
      }), instances.current.size - 1 === idx && onHalt(ctrl));
    };

    for (var _iterator = instances.current.entries(), _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
      var _ref3;

      var _ret = _loop();

      if (_ret === "break") break;
    }

    requestFrame(function () {
      return props.reset && forceUpdate();
    });
  }, [onRest]);
  /** must hoooks always return something? */

  React.useEffect(function () {
    mounted.current = true;
    return function () {
      return void (mounted.current = false);
    };
  }, []);
  React.useLayoutEffect(function () {
    prevItems.current = items;
    if (updatePropsOnRerender) update(props);
  });
  return [array.map(function (item, idx) {
    return {
      item: item,
      props: instances.current.get(reverse ? instances.current.size - (idx + 1) : idx).getValues()
    };
  }), function (props) {
    return update(props);
  }, function (finished) {
    if (finished === void 0) {
      finished = false;
    }

    for (var _iterator2 = instances.current.entries(), _isArray2 = Array.isArray(_iterator2), _i2 = 0, _iterator2 = _isArray2 ? _iterator2 : _iterator2[Symbol.iterator]();;) {
      var _ref5;

      if (_isArray2) {
        if (_i2 >= _iterator2.length) break;
        _ref5 = _iterator2[_i2++];
      } else {
        _i2 = _iterator2.next();
        if (_i2.done) break;
        _ref5 = _i2.value;
      }

      var _ref6 = _ref5,
          ctrl = _ref6[1];
      ctrl.stop(finished);
    }
  }];
}

function parseKeyframedUpdate(slots, config, f, setNext, cancel) {
  if (Array.isArray(slots)) {
    var q = Promise.resolve();

    var _loop = function _loop(i) {
      var slot = slots[i];
      var last = i === slots.length - 1;
      q = q.then(function () {
        return setNext(_extends({
          config: config && (Array.isArray(config) ? config[i] : config)
        }, f(slot)), last);
      });
    };

    for (var i = 0; i < slots.length; i++) {
      _loop(i);
    }
  } else if (typeof slots === 'function') {
    var index = 0;
    slots(function (animatedProps, last) {
      if (last === void 0) {
        last = false;
      }

      var props = _extends({
        config: config && Array.isArray(config) ? config[index++] : config
      }, f(animatedProps));

      return setNext(props, last);
    }, function (finished) {
      if (finished === void 0) {
        finished = true;
      }

      return cancel && cancel(finished);
    });
  } else {
    setNext(_extends({
      config: config && Array.isArray(config) ? config[0] : config
    }, f(slots)), true);
  }
}
function onKeyframesHalt(_ref) {
  var resolverRef = _ref.resolverRef,
      lastRef = _ref.lastRef,
      mounted = _ref.mounted,
      onRestRef = _ref.onRestRef;
  return function (ctrl) {
    return function (_ref2) {
      var finished = _ref2.finished;
      resolverRef.current && resolverRef.current();
      var merged = ctrl && ctrl.merged;
      var _onRestRef$current = onRestRef.current,
          onRest = _onRestRef$current.onRest,
          onKeyframeRest = _onRestRef$current.onKeyframeRest;

      if (finished && lastRef.current && mounted.current) {
        onKeyframeRest && onKeyframeRest(merged);
        onRest && onRest(merged);
      }
    };
  };
}
function setNext(resolverRef, lastRef, setAnimation, onRestRef, shouldForceUpdateRef, forceUpdate) {
  if (resolverRef === void 0) {
    resolverRef = {
      current: null
    };
  }

  if (lastRef === void 0) {
    lastRef = {
      current: null
    };
  }

  if (onRestRef === void 0) {
    onRestRef = {
      current: {
        onKeyframeRest: null
      }
    };
  }

  if (shouldForceUpdateRef === void 0) {
    shouldForceUpdateRef = {
      current: false
    };
  }

  return function (props, last) {
    return new Promise(function (resolve) {
      var _ref3 = props || {},
          onRest = _ref3.onRest,
          animatedProps = _objectWithoutPropertiesLoose(_ref3, ["onRest"]);

      onRestRef.current.onKeyframeRest = onRest;
      resolverRef.current = resolve;
      lastRef.current = last;
      setAnimation && setAnimation(animatedProps);
      shouldForceUpdateRef.current && requestFrame(function () {
        shouldForceUpdateRef.current = false;
        forceUpdate();
      });
    });
  };
}
/**
 *
 * @param {(useSpring | useTrail)} useImpl
 * @param {Object} props
 * @param {Array=} props.items // only needed when using Trail primitive
 * @param {Object} props.states
 * @param {Function} props.filter
 * @param {SpringProps} ...props
 * @param {String} state
 * @param {SpringProps} initialProps
 */

var useKeyframesImpl = function useKeyframesImpl(useImpl) {
  return function (props, initialProps) {
    if (initialProps === void 0) {
      initialProps = null;
    }

    return function (state) {
      if (state === void 0) {
        state = '__default';
      }

      var resolverRef = React.useRef(null);
      var onRestRef = React.useRef({
        onRest: null,
        onKeyframeRest: null
      });
      var lastRef = React.useRef(true);
      var mounted = React.useRef(false); // need to force a rerender for when the
      // animated controller has finally accepted
      // some props

      var _React$useState = React.useState(),
          forceUpdate = _React$useState[1];

      var shouldForceUpdateRef = React.useRef(!initialProps);

      var _ref4 = function () {
        if (Array.isArray(props) || typeof props === 'function') {
          var _states;

          return {
            states: (_states = {}, _states[state] = props, _states)
          };
        } else {
          var onRest = props.onRest,
              _config = props.config,
              _filter = props.filter,
              rest = _objectWithoutPropertiesLoose(props, ["onRest", "config", "filter"]);

          onRestRef.current.onRest = onRest;
          return {
            states: rest,
            config: _config,
            filter: _filter
          };
        }
      }(),
          states = _ref4.states,
          config = _ref4.config,
          _ref4$filter = _ref4.filter,
          filter = _ref4$filter === void 0 ? function (states) {
        return states;
      } : _ref4$filter;

      var _useImpl = useImpl(_extends({}, initialProps, {
        onKeyframesHalt: onKeyframesHalt({
          resolverRef: resolverRef,
          lastRef: lastRef,
          mounted: mounted,
          onRestRef: onRestRef
        }),
        updatePropsOnRerender: false
      })),
          animProps = _useImpl[0],
          setAnimation = _useImpl[1],
          cancel = _useImpl[2];

      var setNextKeyFrame = setNext(resolverRef, lastRef, setAnimation, onRestRef, shouldForceUpdateRef, forceUpdate);
      React.useEffect(function () {
        mounted.current = true;
        return function () {
          return mounted.current = false;
        };
      }, []);
      React.useEffect(function () {
        void (mounted.current && parseKeyframedUpdate(states[state], callProp(config, state), filter, setNextKeyFrame, cancel));
      }, [state]);
      return shouldForceUpdateRef.current && Array.isArray(animProps) ? [] : animProps;
    };
  };
};

var useKeyframes = {
  spring: function spring(arg) {
    return useKeyframesImpl(useSpring)(arg);
  },
  trail: function trail(arg) {
    return useKeyframesImpl(useTrail)(arg);
  }
};

var guid = 0;

var get = function get(props) {
  var items = props.items,
      _props$keys = props.keys,
      keys = _props$keys === void 0 ? function (states) {
    return states;
  } : _props$keys,
      rest = _objectWithoutPropertiesLoose(props, ["items", "keys"]);

  items = toArray(items !== void 0 ? items : null);
  keys = typeof keys === 'function' ? items.map(keys) : toArray(keys); // Make sure numeric keys are interpreted as Strings (5 !== "5")

  return _extends({
    items: items,
    keys: keys
  }, rest);
};

function debounce(func, delay) {
  if (delay === void 0) {
    delay = 0;
  }

  var timeoutId;
  return function () {
    var context = this;
    var args = arguments;
    clearTimeout(timeoutId);
    timeoutId = setTimeout(function () {
      return func.apply(context, args);
    }, delay);
  };
} // function bufferedCall ()


function calculateDiffInItems(_ref, props) {
  var prevProps = _ref.prevProps,
      state = _objectWithoutPropertiesLoose(_ref, ["prevProps"]);

  var _get = get(prevProps || {}),
      _keys = _get.keys,
      _items = _get.items;

  var _get2 = get(props),
      keys = _get2.keys,
      items = _get2.items,
      unique = _get2.unique,
      trail = _get2.trail,
      update = _get2.update,
      enter = _get2.enter,
      leave = _get2.leave,
      config = _get2.config;

  var currSet = new Set(keys);
  var prevSet = new Set(_keys);
  var deleted = state.deleted.concat();

  var current = _extends({}, state.current);

  var removed = state.transitions.filter(function (item) {
    return !item.destroyed && !currSet.has(item.originalKey);
  });
  var added = keys.filter(function (key) {
    return !prevSet.has(key);
  });

  var updated = _keys.filter(function (key) {
    return currSet.has(key);
  });

  var delay = 0;
  added.forEach(function (key) {
    var keyIndex = keys.indexOf(key);
    var item = items[keyIndex];
    var state = 'enter';

    if (unique && deleted.find(function (d) {
      return d.originalKey === key;
    })) {
      deleted = deleted.filter(function (t) {
        return t.originalKey !== key;
      });
    }

    current[key] = {
      item: item,
      state: state,
      trail: delay = delay + trail,
      key: unique ? String(key) : guid++,
      originalKey: key,
      destroyed: false,
      config: callProp(config, item, state),
      to: callProp(enter, item)
    };
  });
  removed.forEach(function (_ref2) {
    var item = _ref2.item,
        rest = _objectWithoutPropertiesLoose(_ref2, ["item"]);

    var keyIndex = _keys.indexOf(item.originalKey);

    var state = 'leave';
    deleted.unshift(_extends({}, rest, {
      item: item,
      state: state,
      left: _keys[Math.max(0, keyIndex - 1)],
      right: _keys[Math.min(_keys.length, keyIndex + 1)],
      destroyed: true,
      trail: delay = delay + trail,
      config: callProp(config, item, state),
      to: callProp(leave, item)
    }));
    delete current[item.originalKey];
  });
  updated.forEach(function (key) {
    var keyIndex = keys.indexOf(key);
    var item = items[keyIndex];
    var state = 'update';
    current[key] = _extends({}, current[key], {
      item: item,
      state: state,
      destroyed: false,
      trail: delay = delay + trail,
      config: callProp(config, item, state),
      to: callProp(update, item)
    });
  });
  var transitions = keys.map(function (key) {
    return current[key];
  }); // this is so the latest deleted item might find its position first
  // as older deleted items might reference later deleted items to their left or right

  deleted.forEach(function (_ref3) {
    var left = _ref3.left,
        right = _ref3.right,
        item = _objectWithoutPropertiesLoose(_ref3, ["left", "right"]);

    var pos; // Was it the element on the left, if yes, move there ...

    if ((pos = transitions.findIndex(function (t) {
      return t.originalKey === left;
    })) !== -1) {
      pos += 1;
    } // Or how about the element on the right ...


    if (pos === -1) pos = transitions.findIndex(function (t) {
      return t.originalKey === right;
    }); // And if nothing else helps, move it to the start \_()_/

    pos = Math.max(0, pos);
    transitions = transitions.slice(0, pos).concat([item], transitions.slice(pos));
  });
  return {
    deleted: deleted,
    updated: updated,
    current: current,
    transitions: transitions
  };
}

var removeDeleted = function () {
  var deleted = [];
  var debounceUpdateState = debounce(function (_ref4, setState) {
    var state = _ref4.current;
    setState(_extends({}, state, {
      deleted: state.deleted.filter(function (item) {
        return deleted.findIndex(function (val) {
          return val.originalKey === item.originalKey;
        }) === -1;
      })
    }));
    deleted = [];
  }, 200);
  return function (key, stateRef, setState) {
    deleted.push(key);
    debounceUpdateState(stateRef, setState);
  };
}();
/**
 * @param {TransitionProps} props
 */


function useTransition(props) {
  var _get3 = get(props),
      items = _get3.items,
      _currentKeys = _get3.keys,
      trail = _get3.trail,
      from = _get3.from,
      initial = _get3.initial,
      onRest = _get3.onRest,
      onDestroyed = _get3.onDestroyed,
      unique = _get3.unique,
      _get3$filter = _get3.filter,
      filter = _get3$filter === void 0 ? function (states) {
    return states;
  } : _get3$filter,
      rest = _objectWithoutPropertiesLoose(_get3, ["items", "keys", "trail", "from", "initial", "onRest", "onDestroyed", "unique", "filter"]);

  var instances = React.useRef(new Map([]));
  var first = React.useRef(true);
  var mounted = React.useRef(false);
  var activeSlots = React.useRef({});

  var _React$useState = React.useState({
    deleted: [],
    current: {},
    transitions: [],
    prevProps: null
  }),
      state = _React$useState[0],
      setState = _React$useState[1]; // keep current state in a mutable ref so
  // promise callback will always have access to latest state


  var stateRef = React.useRef(state);
  stateRef.current = state;
  var memoizedDiffInItemsCalc = React.useMemo(function () {
    return calculateDiffInItems(state, props);
  }, [props.items, state.deleted]);
  React.useEffect(function () {
    mounted.current = true;
    return function () {
      return void (mounted.current = false);
    };
  }, []); // Prop changes effect

  React.useEffect(function () {
    var deleted = memoizedDiffInItemsCalc.deleted,
        current = memoizedDiffInItemsCalc.current,
        transitions = memoizedDiffInItemsCalc.transitions;
    transitions.forEach(function (_ref5) {
      var slot = _ref5.state,
          to = _ref5.to,
          config = _ref5.config,
          key = _ref5.key,
          originalKey = _ref5.originalKey,
          item = _ref5.item,
          destroyed = _ref5.destroyed;
      !instances.current.has(key) && instances.current.set(key, {
        ctrl: new Controller(_extends({}, from || {}, first.current && initial || {}, {
          config: config
        })),
        resolve: {
          current: null
        },
        last: {
          current: true
        }
      });
      var instance = instances.current.get(key);
      var ctrl = instance.ctrl,
          resolve = instance.resolve,
          last = instance.last;

      if (slot === 'update' || slot !== activeSlots.current[key]) {
        var onEnd = function onEnd(_ref6) {
          var finished = _ref6.finished;
          resolve.current && resolve.current(ctrl.merged);

          if (last.current && mounted.current && finished) {
            destroyed && onDestroyed && onDestroyed(item);
            destroyed && removeDeleted(originalKey, stateRef, setState);
            onRest && onRest(item, slot, ctrl.merged);
          }
        };

        // add the current running slot to the active slots ref so the same slot isnt re-applied
        activeSlots.current[key] = slot;

        var updater = function updater(props) {
          ctrl.update(props, onEnd);
        };

        var setNextKeyframe = setNext(resolve, last, updater);
        parseKeyframedUpdate(to, config, filter, setNextKeyframe, function (finished) {
          if (finished === void 0) {
            finished = true;
          }

          return ctrl.stop(finished);
        });
      }
    });
    first.current = false;
    setState(_extends({}, state, {
      deleted: deleted,
      current: current,
      transitions: transitions,
      prevProps: props
    }));
  }, [props.items]);
  return state.transitions.map(function (_ref7) {
    var item = _ref7.item,
        state = _ref7.state,
        key = _ref7.key;
    return {
      item: item,
      key: key,
      state: state,
      props: instances.current.get(key).ctrl.getValues()
    };
  });
}

var domElements = ['a', 'abbr', 'address', 'area', 'article', 'aside', 'audio', 'b', 'base', 'bdi', 'bdo', 'big', 'blockquote', 'body', 'br', 'button', 'canvas', 'caption', 'cite', 'code', 'col', 'colgroup', 'data', 'datalist', 'dd', 'del', 'details', 'dfn', 'dialog', 'div', 'dl', 'dt', 'em', 'embed', 'fieldset', 'figcaption', 'figure', 'footer', 'form', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'head', 'header', 'hgroup', 'hr', 'html', 'i', 'iframe', 'img', 'input', 'ins', 'kbd', 'keygen', 'label', 'legend', 'li', 'link', 'main', 'map', 'mark', 'marquee', 'menu', 'menuitem', 'meta', 'meter', 'nav', 'noscript', 'object', 'ol', 'optgroup', 'option', 'output', 'p', 'param', 'picture', 'pre', 'progress', 'q', 'rp', 'rt', 'ruby', 's', 'samp', 'script', 'section', 'select', 'small', 'source', 'span', 'strong', 'style', 'sub', 'summary', 'sup', 'table', 'tbody', 'td', 'textarea', 'tfoot', 'th', 'thead', 'time', 'title', 'tr', 'track', 'u', 'ul', 'var', 'video', 'wbr', // SVG
'circle', 'clipPath', 'defs', 'ellipse', 'foreignObject', 'g', 'image', 'line', 'linearGradient', 'mask', 'path', 'pattern', 'polygon', 'polyline', 'radialGradient', 'rect', 'stop', 'svg', 'text', 'tspan'];
var extendedAnimated = domElements.reduce(function (acc, element) {
  acc[element] = createAnimatedComponent(element);
  return acc;
}, createAnimatedComponent);

export { config, extendedAnimated as animated, interpolate$1 as interpolate, Globals, useSpring, useTrail, useTransition, useKeyframes };
